// @ali/ldmap/link_ui/link_process.js
'use strict';
var EventEmitter = require('./../libs/event');
var Util = require('./../core/utils');
var $ = require('jquery');
var NodeBase = require('./node_base');
/**
 * ## LinkProcess 连接器 专门解析函数
 * @class Node
 */

function LinkProcess(options) {
  this.initialize(options);
}

EventEmitter.extend(LinkProcess, {
  options: {
    type: 'data'
  },
  initialize: function(options) {
    this.options = Util.deepMerge(this.options, options);
  },
  linkTo: function(component) {
    this.component = component;
  },
  process: function (d) {
    var from = this.from = d.from, to = this.to = d.to, link = this.link = d.link;
    if(this.check(from, to, link)){
      this.updateUI();
    }
  },
  check: function(from, to, link) {
    if(from.getType() === to.getType()){
      link.destory();
      return;
    } 
    return true;
  },
  updateUI: function(){
    var from = this.from, to = this.to, link = this.link;
    var fromPos = from.getPosition();
    var toPos = to.getPosition();
    if(fromPos.left.left > toPos.right.left){
      var tmp = from;
      this.from = from = to;
      this.to = to = tmp;
      link.modify(toPos.right.left, toPos.right.top, fromPos.left.left, fromPos.left.top);
    } else{
      link.modify(fromPos.right.left, fromPos.right.top, toPos.left.left, toPos.left.top);
    }
  },
  initEvents: function() {
  }
});

module.exports = LinkProcess;
