// @ali/common-utils/sprite/index.js
'use strict';
var gradient = require('./gradient');
var Utils = require('bcore/utils');

function sprite(type, options){
  type = type || 'gradient';
  options = options || {};
  var canvas = document.createElement('canvas');
  canvas.width = options.width || options.w || options.size || options.radius * 2 || 80;
  canvas.height = options.height || options.h || options.size || options.radius * 2 || 80;
  var ctx = canvas.getContext('2d');
  if(typeof(type) === 'string'){
    switch (type){
      case 'gradient':
        var defaults = {
          isContinue: false,
          type: 'radian',
          color:{
            from: 'rgba(0,150,150,0)',
            to: 'rgba(250,250,250,1)',
            space: 'rgb',
            easing: 'linear.None.2'
          }
        };
        console.log(defaults, options);
        options = Utils.deepMerge(defaults, options);
        return gradient(ctx, options).getCanvas();
        break;
    }
  }
}

module.exports = {
  'gradient': gradient,
  'sprite': sprite
};
