// @ali/d-chart/src/theme/country/countryNumber.js
var DataV = require('../../charts/number/number');
var _ = require('../../util');
var d3 = require('../../libs/d3');

function CountryNumber(options) {
  this.opt = {
    margin: {left: 0, top: 30, right: 0, bottom: 30},
    axis : {
      orient : 'bottom'
    }
  };
  this.opt = _.extend({}, this.opt, options);
  this.blockWidth = 142;
  this._init(this.opt);
  this._initLayout();
}

DataV.extend(CountryNumber, {
  _afterRender : function () {

    var bw = this.blockWidth;

    var svg = this.svg;
    var text = svg.selectAll('.number')
      .each(function () {
        var that = d3.select(this);
        var w = this.getBBox().width;
        var x0 = parseInt(that.attr('x')) - 6;
        var y0 = parseInt(that.attr('y')) - 20;
        var x1 = x0 + bw;
        var y1 = y0 + 80;
        that.attr('dx', ((bw - w) / 2) + 'px');
        var t1 = 2;
        var t2 = 30;
        svg.append('path')
          .attr({
            'class' : 'number-bg',
            'd' : function () {
              return 'M' + (x0 + t1) + ' ' + y1 +
                'T' + x0 + ' ' + (y1 - t1) +
                'L' + x0 + ' ' + (y0 + t2) +
                'T' + (x0 + t2) + ' ' + y0 +
                'L' + (x1 - t1) + ' ' + y0 +
                'T' + x1 + ' ' + (y0 + t1) +
                'L' + x1 + ' ' + (y1 - t2) +
                'T' + (x1 - t2) + ' ' + y1 +
                'Z';
            }
          })
      });

    var axis = this._com.getComs('axis');
    var arrX = axis[0].selectAll('.tick')
      .each(function () {
        var that = d3.select(this);
        var x0 = - bw / 2;
        var y0 = -5;
        var x1 = bw / 2;
        var y1 = 25;
        var gap = 10;
        that.append('polygon')
          .attr({
            'class' : 'axis-bg',
            'points' : function () {
              var pos = [
                {x:x0, y:y1},
                {x:x0, y:y0},
                {x:x1-gap*2, y:y0},
                {x:x1, y:y0+gap},
//                {x:x1, y:y0},
                {x:x1, y:y1}
              ];
              return _.pos2Str(pos);
            }
          })
      });

  }
});

module.exports = CountryNumber;
