// @ali/d-chart/src/theme/country/countryLine.js
var DataV = require('../../charts/line/lineMulti');
var _ = require('../../util');
var d3 = require('../../libs/d3');

function CountryLine(options) {
  this.opt = {
    margin : {left: 60, top:10, right:40, bottom:20},
    xaxis : {
      min : new Date('2009-08-11'),
      max : new Date('2015-05-23'),
      type : 'time',
//      format : '%Y'
    },
    yaxis : {
      max : 10,
      min : 0
    },
    beforeLayout : {
      legend : {
        position : 'float',
        height : '54',
        width : '84'
      }
    },
    legend : {
      type : 'vertical',
      customRender : function (label, pos, i) {
        var x = pos.x;
        var y = pos.y;
        var h = pos.height;
        this.append('circle')
          .attr({
            'class' : 'back',
            'cx' : x + 15,
            'cy' : y + 22,
            'r' : 10
          });

        this.append('circle')
          .attr({
            'class' : 'front',
            'cx' : x + 15,
            'cy' : y + 22,
            'r' : 3
          });

        this.append('text')
          .attr({
            'class' : 'label-text',
            'x' : x + 30,
            'y' : y + h - 3
          }).text(label);
      }
    }
  };
  this.opt = _.extend({}, this.opt, options);
  this._init(this.opt);
  this._initLayout();
}

DataV.extend(CountryLine, {
  _beforeRender : function () {
    var bg = this.renderBG();
    bg.renderGrid(13, 6);
  },
  _afterRender : function () {
    var opt = this.opt;
    var axiss = this._com.getComs('axis');

    var xAxis = axiss[0].getX();
    var yAxis = axiss[1].getX();

    this.svg.selectAll('.lines')
      .each(function () {
        var lines = d3.select(this);
        var data = lines.data();
        data[0].length && data[0].forEach(function (obj) {
          var x = obj[opt.x];
          obj[opt.y].forEach(function (y, i) {

            //加2个同心圆
            var point = lines.append('g')
              .attr('class', 'point point' + i);

            point.append('circle')
              .attr({
                'class' : 'back',
                'cx' : xAxis(x),
                'cy' : yAxis(y),
                'r' : 10
              });

            point.append('circle')
              .attr({
                'class' : 'front',
                'cx' : xAxis(x),
                'cy' : yAxis(y),
                'r' : 3
              });
          });
        });
      });

    var arrX = axiss[0].selectAll('.tick')[0];
    var arrY = axiss[1].selectAll('.tick')[0];
    d3.select(arrX[arrX.length - 1]).select('text')
      .text('年份');
    d3.select(arrY[arrY.length - 1]).select('text')
      .text('贸易额');
    var legend = this.renderLegend(opt.legend);
  }
});

module.exports = CountryLine;
