// @ali/d-chart/src/theme/country/countryBarHori.js
var DataV = require('../../charts/bar/barMultiHori');
var _ = require('../../util');
var d3 = require('../../libs/d3');
var textures = require('textures');

function CountryBarHori(options) {
  this.opt = {
    margin : {left: 100, top: -37, right: 0, bottom: 0},
    xaxis : {
      padding : 0.3,
      min : 0,
      max : 9
    },
    yaxis : {
      type : 'category',
      padding : 0.6
    }
  };
  this.opt = _.extend({}, this.opt, options);
  this._init(this.opt);
  this._initLayout();
}

DataV.extend(CountryBarHori, {
  _afterRender : function () {
    var axis = this._com.getComs('axis')[1];
    axis.attr({
      'transform' : 'translate(' + (-40) + ', 0)'
    });
    axis.append('polygon', '.tick');
    axis.selectAll('polygon')
      .attr({
        'points' : function () {
          var w = 12;
          var s = 6;
          var points = [
            {x: s, y: 0 },
            {x: s + w, y: - w*2},
            {x: s + w + w, y:  -w*2},
            {x: s + w + w, y:  w*2},
            {x: s + w, y: w*2},
          ];
          return _.pos2Str(points);
        }
      });
    this.svg.selectAll('.bars')
      .each(function () {
        var bars = d3.select(this);
        bars.selectAll('.bar')
          .each(function () {
            var bar = d3.select(this);
            var x = parseInt(bar.attr('x'));
            var y = parseInt(bar.attr('y'));
            var height = parseInt(bar.attr('height'));
            var index = bar.classed('y1') ? 0 : 1;
            bars.insert('text')
              .attr({
                'x' : x + 4,
                'y' : y + height - 3
              }).text(function () {
                var args = ['村点数', '淘村数'];
                return args[index];
              });
          });
      });
  }
});

module.exports = CountryBarHori;
