// @ali/ldmap/render/canvas.js
var L = require('leaflet');
var $ = require('jquery');
var dmap = L.dmap = L.dmap || {};

var requestAnimationFrame =
  window.requestAnimationFrame ||
  window.mozRequestAnimationFrame ||
  function(callback) {
    return window.setTimeout(callback, 1000 / 60);
  };
var cancelAnimationFrame =
  window.cancelAnimationFrame ||
  window.cancelRequestAnimationFrame ||
  window.mozCancelAnimationFrame ||
  function(id) {
    window.clearTimeout(id);
  };

function Canvas(map, node, opt) {
  opt = opt || {};
  this.map = map;
  this.opt = opt;
  this.init(node, opt);
  if (opt.darken) this.setDarken(opt.darken);
  this.styles(opt);
  this.onDefaults();
}

/**
 * init 创建一个和传入节点等大的canvas
 * @param  {Object} node jquery节点对象
 * @param  {Object} opt  [description]
 */
Canvas.prototype.init = function (node, opt) {
  var map = this.map;
  if(typeof(node) === 'string'){
    var nodes = map.getPanes();
    if(nodes[node]){
      node = nodes[node];
      var mapSize = map.getSize();
      opt.width = mapSize.x;
      opt.height = mapSize.y;
      this.isMapPane = 1;
    }
  }
  node = $(node);
  var w = this.w = opt.width || node.width();
  var h = this.h = opt.height || node.height();
  var canvas;
  if (node.is('div')) { //如果是div 创建一个canvas
    var quality = this.quality = opt.quality || 1.5;
    canvas = $('<canvas width="' + (w * quality) + '" height="' + (h * quality) + '"><canvas>')
      .css({
        'position': 'absolute',
        'left': '0',
        'top': '0',
        'pointerEvents': 'none',
        'width': w + 'px',
        'height': h + 'px'
      });
    node.append(canvas);
    canvas = this.canvas = canvas[0];
    var ctx = this.ctx = canvas.getContext('2d');
    ctx.scale(quality, quality);
    return;
  }

  if (node.is('canvas')) { //如果是canvas 修改
    canvas = this.canvas = node[0];
    this.ctx = canvas.getContext('2d');
  }
};

/**
 * styles 对canvas风格上的一些设置 如叠加模式
 * @param  {Object} cfg 配置
 */
Canvas.prototype.styles = function (cfg) {
  cfg = cfg || {};
  var lighter = cfg.lighter;
  var ctx = this.ctx;
  lighter && ctx && (ctx.globalCompositeOperation = 'lighter'); //是否叠加变量
  var opacity = cfg.opacity;
  (opacity !== undefined) && (opacity !== null) && ctx && (ctx.globalAlpha = opacity); //是否有透明度
};

/**
 * onDefaults 默认地图事件产生的操作
 */
Canvas.prototype.onDefaults = function() {
  var clear = this.clear.bind(this);
  var resetPos = this.resetPos.bind(this);
  this.map
    .on('zoomstart', clear)
    .on('zoomend', resetPos)
    .on('movestart', clear)
    .on('resize', function(){
      clear();
      resetPos();
    })
    .on('moveend', resetPos);
};

/**
 * bind 和地图事件产生绑定
 * @param  {String}   name 事件名称
 * @param  {Function} cb   绑定的
 * @return {[type]}        [description]
 */
Canvas.prototype.bind = function (name, hook) {
  this[name] && this.map.off(name, this[name].bind(this));
  this[name] = hook;
  this.map.on(name, hook.bind(this));
};

/**
 * onZoomstart 给地图缩放开始事件绑定函数
 * @param  {Function} hook 函数
 */
Canvas.prototype.onZoomstart = function (hook) {
  this.bind('zoomstart', hook);
};

/**
 * onMovestart 给地图移动开始事件绑定函数
 * @param  {Function} hook 函数
 */
Canvas.prototype.onMovestart = function (hook) {
  this.bind('movestart', hook);
};

/**
 * onMoveend 给地图移动结束事件绑定函数
 * @param  {Function} hook 函数
 */
Canvas.prototype.onMoveend = function (hook) {
  this.bind('moveend', hook);
};

/**
 * onZoomend 给地图缩放结束事件绑定函数
 * @param  {Function} hook 函数
 */
Canvas.prototype.onZoomend = function (hook) {
  this.bind('zoomend', hook);
};

/**
 * onUpdate 给地图变化事件结束事件绑定函数
 * @param  {Function} hook 函数
 */
Canvas.prototype.onUpdate = function (hook) {
  this.bind('moveend', hook);
  this.bind('resize', hook);
};

Canvas.prototype.setDarken = function (darken) {
  if(darken && darken.alpha){
    this._darkenAlphaK = darken.alpha;
    this.cancelDarken();
    this.darken();
  }
};

Canvas.prototype.cancelDarken = function () {
  cancelAnimationFrame(this.darkenLoopId);
};

/**
 * darken 让一个canvas 所有的像素
 * @param  {String} alphaK 让本canvas所有像素的alpha值乘以系数，就是变透明
 */
Canvas.prototype.darken = function () {
  var alphaK = this._darkenAlphaK;
  var ctx = this.ctx;
  if (alphaK !== null && alphaK !== undefined) {
    if (alphaK >= 0) {
      var imageData = this.ctx.getImageData(0, 0, this.w, this.h);
      var data = imageData.data;
      for (var k = 3; k < data.length; k += 4) { ///对ctx暗化,每次对每个像素的alpha值乘以系数
        data[k] = data[k] * alphaK;
      }
      ctx.putImageData(imageData, 0, 0);
    } else {
      this.clear();
    }
  }
  this.darkenLoopId = requestAnimationFrame(this.darken.bind(this));
};

/**
 * resetPos 重新排列节点位置
 */
Canvas.prototype.resetPos = function() {
  var canvas = this.canvas;
  if (this.isMapPane) {
    var domPosition = L.DomUtil.getPosition(this.map.getPanes().mapPane);
    if (domPosition) {
      L.DomUtil.setPosition(canvas, {
        x: -domPosition.x,
        y: -domPosition.y
      });
    }
  }
};
/**
 * clear 清除目前的绘制
 */
Canvas.prototype.clear = function () {
  this.ctx.clearRect(0, 0, this.w, this.h);
};

/**
 * destroy 销毁节点
 */
Canvas.prototype.destroy = function () {
  this.canvas.remove();
};

/**
 * pt  在canvas上根据传入图片画点图
 * @param  {Object} img 图片/canvas格式的sprite小图片
 * @param  {Float} x   中心点的X坐标
 * @param  {Float} y   中心点的X坐标
 * @param  {Float} w   图片宽度
 * @param  {Float} h   图片长度
 */
Canvas.prototype.pt = function (img, x, y, w, h) {
  h = h || w;
  this.ctx.drawImage(img, x - w / 2, y - h / 2, w, h);
};

Canvas.prototype.test = function (cfg) {
  this.ctx.fillStyle = '#900';
  this.ctx.fillRect(20, 202, 333 + Math.random() * 100, 33);
  this.ctx.fillRect(24, 223, 33, 333);
};

dmap.Canvas = Canvas;
dmap.canvas = function (map, canvas, opt) {
  return new Canvas(map, canvas, opt);
};
module.exports = Canvas;
