// @ali/gaode_map_service/test/demo/data_layer_marker.js
'use strict';

var $ = require('jquery');
var _ = require('lodash');

var L = require('leaflet');
require('./../layer/anim_marker');
var animMarker = L.dmap.animMarker;
require('./../core/utils');

var StreamLayer = require('./../data_layer/stream_layer');

function Demo(container) {
  container = container || $('#map-container');
  this.container = container;
  this.initMap();
  this.initVisual();
  this.list = {};
  this.loop();
}

Demo.prototype.initVisual = function () {
  var map = this.map;
  this.dataLayer = new StreamLayer({
    visualize: function () {
      var aMarker = animMarker({
        map: map,
        className: 'rect',
        animClass: 'fadeInOut',
        life: Math.random() * 12 + 1
      });
      aMarker
      .on('mouseover', function () {
        this.pause();
      })
      .on('mouseout', function () {
        this.resume();
      });
      return aMarker;
    },
    map: map
  });
};

Demo.prototype.initMap = function() {
  var map = this.map = L.map(this.container.attr('id'), {
    'center': L.latLng(39.90973623, 116.34521484),
    'zoom': 10,
    'attributionControl': false,
    'crs': L.CRS.EPSG4326
  });
  L.tileLayer('http://t.mapabc.com/maptile?t=3&x={x}&y={y}&z={z}').addTo(map);
};

function randomPick(list) { //随机选择数据
  if (typeof(list) === 'object') {
    list = _.values(list);
  }
  return list[Math.floor(Math.random() * list.length)];
}

Demo.prototype.mock = function () {
  return {
    pos: {
      'lng': 116.34521484 + (Math.random() - 0.5) * 1,
      'lat': 39.90973623 + (Math.random() - 0.5) * 1
    },
    'text': 'text_' + parseInt(Math.random() * 30),
    'id': 'id_' + parseInt(Math.random() * 100)
  };
};

Demo.prototype.mocks = function () {
  var result = {};
  for (var k = 0; k < 3 * Math.random(); k++) {
    var d = this.mock();
    var id = d.id;
    result[id] = d;
  }
  return result;
};

Demo.prototype.loop = function () {
  if (Math.random() < 0.9) {
    var datas = this.mocks();
    this.dataLayer.data(datas);
  }
  window.requestAnimationFrame(this.loop.bind(this));
};

module.exports = Demo;
