// @ali/gaode_map_service/test/demo/flyingLineSvg_demo.js
var $ = require('jquery');
var L = require('leaflet');
require('./../layer/flyingLineGradient');
var flyingLineGradient = L.dmap.flyingLineGradient;
var demo2d = require('./demo');

var datas = [{
  from: {
    lng: 97.34652,
    lat: 33.819569
  },
  to: {
    lng: 107.61673,
    lat: 43.901106
  },
  phone: 11111
}, {
  from: {
    lng: 103.34652,
    lat: 33.819569
  },
  to: {
    lng: 107.61673,
    lat: 45.901106
  },
  phone: 22222
}];

function Test(container) {
  container = container || $('#map-container');
  this.map = demo2d(container);
  this.flyingLines = {};
  this.render();
}

Test.prototype.render = function () {
  var map = this.map, line, data;
  for (var i = 0; i < datas.length; i++) {
    data = datas[i];
    line = flyingLineGradient({//创建飞线
      'list': this.flyingLines,
      'weight': 2,
      'lifeEnd': 'loop',
      'range': 1,
      'colorNormal': 'rgba(255,0,0,0.3)',
      'interactiveLine': 1,
      'map': map
    });
    line.data(data);
  }
};

module.exports = Test;
