// @ali/gaode_map_service/test/demo/scatter_canvas_demo.js
/*
jquery leaflet
 */

var $ = require('jquery');
var L = require('leaflet');
require('./../layer/scatter_canvas');
require('./tilelayer');
var Utils = require('./../core/utils');
var dotData = require('./data/peopleBeijin');

//dmap的域
var dmap = L.dmap;
// var blending = require('@ali/common-utils/color/blending');

function scatterDemo(container) {
  var ctx = $('<canvas width="80" height="80"></canvas>')[0].getContext('2d');
  var colorFunc = blending('rgba(245,230,0,0.1842506350550381)', 'rgba(255,0,0,0)', {
    easing: 'Linear.In.1.7341236240474176',
    type: 'rgb'
  }).toFunc();
  var sprite = Utils.gradient(ctx, {
    isContinue: false,
    color: colorFunc
  });

  //建立地图
  var map = L.dmap.map(container.attr('map-container'), {
    'center': L.latLng(39.289492, 118.956548),
    'zoom': 4,
    'attributionControl': false
  });
  //调用atlas地图瓦片服务
  dmap.tileLayer('AtlasDark').addTo(map);

  scatterCanvas = dmap.scatterCanvas(map, {
    container: $('#canvas-container')
  });
  scatterCanvas.data(dotData);
}

module.exports = scatterDemo;
