// @ali/d-chart/src/instance/render.js
var exports = module.exports;
var _ = require('../util');

exports._beforeRender = function () {
  this.emit('before.render');
};

exports._afterRender = function () {
  this.emit('after.render');
};

exports.renderCom = function (type, options) {
  var opt = _.extend({}, this.$options, {
    parentEl : this.$el,
    svg : this.svg,
    _data : this._data,
  }, options);
  var com = this._com.add(type, opt);
  com.render();
  return com;
};

exports.renderAxis = function (options) {
  var opt = _.extend({}, this.$options, {
    svg : this.svg,
    _data : this._data,
  }, options);
  var axis = this._com.addAxis(opt);
  axis.render();
  axis.addClass('axis' + this._com.getAccountByType('axis'));
  return axis;
};

exports.renderBG = function (options) {
  var opt = _.extend({}, this.$options, {
    svg : this.svg,
    _data : this._data,
  }, options);
  var bg = this._com.addBG(opt);
  bg.render();
  return bg;
};

exports.renderLegend = function (options) {
  var opt = _.extend({}, this.$options, {
    svg : this.svg,
    _data : this._data,
  }, options);
  var legend = this._com.addLegend(opt);
  legend.render();
  return legend;
};

exports.renderTime = function (options) {
  var opt = _.extend({}, this.$options, {
    svg : this.svg,
    _data : this._data,
  }, options);
  var time = this._com.addTime(opt);
  time.render();
  return time;
};

exports.renderNumber = function (options) {
  var opt = _.extend({}, this.$options, {
    svg : this.svg,
    _data : this._data,
  }, options);
  var num = this._com.addNumber(opt);
  num.render();
  return num;
};

exports.emptyRender = function () {
  this.svg.remove();
};
